projects/packages/jetpack-mu-wpcom/tests/php/features/launchpad/class-launchpad-task-definitions-test.php

```diff
@@ -0,0 +1,114 @@
+					'test_task_with_id_map' => array(
+						'id_map' => 'THIS_IS_THE_MAP',
+					),
+				);
+			}
+		);
+	}
+
+	/**
+	 * Tests wether the wpcom_mark_launchpad_task_complete works correctly.
+	 */
+	public function test_wpcom_mark_launchpad_task_complete() {
+		wpcom_mark_launchpad_task_complete( 'test_task1' );
+		$options = get_option( 'launchpad_checklist_tasks_statuses' );
+		$this->assertTrue( isset( $options['test_task1'] ) );
+		$this->assertTrue( $options['test_task1'] );
+
+		wpcom_mark_launchpad_task_complete(
+			array(
+				'test_task2',
+				'test_task3',
+			)
+		);
+		$options = get_option( 'launchpad_checklist_tasks_statuses' );
+		$this->assertTrue( isset( $options['test_task2'] ) && $options['test_task2'] );
+		$this->assertTrue( isset( $options['test_task3'] ) && $options['test_task3'] );
+
+		$this->assertTrue( count( $options ) === 3 );
+	}
+
+	/**
+	 * Tests wether the wpcom_mark_launchpad_task_incomplete works correctly.
+	 */
+	public function test_wpcom_mark_launchpad_task_incomplete() {
+		wpcom_mark_launchpad_task_incomplete( 'test_task1' );
+		$options = get_option( 'launchpad_checklist_tasks_statuses' );
+		$this->assertTrue( isset( $options['test_task1'] ) && ! $options['test_task1'] );
+
+		wpcom_mark_launchpad_task_incomplete(
+			array(
+				'test_task2',
+				'test_task3',
+			)
+		);
+		$options = get_option( 'launchpad_checklist_tasks_statuses' );
+		$this->assertTrue( isset( $options['test_task2'] ) && ! $options['test_task2'] );
+		$this->assertTrue( isset( $options['test_task3'] ) && ! $options['test_task3'] );
+
+		$this->assertTrue( count( $options ) === 3 );
+	}
+
+	/**
+	 * Tests setting a task as incomplete after it was marked complete.
+	 */
+	public function test_wpcom_launchpad_revert_completed_status() {
+		wpcom_mark_launchpad_task_complete( 'test_task1' );
+		wpcom_mark_launchpad_task_incomplete( 'test_task1' );
+
+		$options = get_option( 'launchpad_checklist_tasks_statuses' );
+		$this->assertTrue( isset( $options['test_task1'] ) && ! $options['test_task1'] );
+	}
+	/**
+	 * Tests wether a correct amount of array elements get stored in the 'launchpad_checklist_tasks_statuses' option.
+	 */
+	public function test_correct_task_count() {
+		wpcom_mark_launchpad_task_complete(
+			array(
+				'test_task2',
+				'test_task3',
+			)
+		);
+
+		$options = get_option( 'launchpad_checklist_tasks_statuses' );
+
+		$this->assertTrue( count( $options ) === 2 );
+	}
+
+	/**
+	 * Tests wether the correct task ID is stored in the 'launchpad_checklist_tasks_statuses' option.
+	 * When tasks have an 'id_map' property, the 'id_map' value should be used as the task ID.
+	 */
+	public function test_correct_task_id_is_stored() {
+		wpcom_mark_launchpad_task_complete( 'test_task_with_id_map' );
+		$options = get_option( 'launchpad_checklist_tasks_statuses' );
+		$this->assertFalse( isset( $options['test_task_with_id_map'] ) );
+		$this->assertTrue( isset( $options['THIS_IS_THE_MAP'] ) );
```

Comment:
I am really confused by this test and why it passes. Based on the following lines in `wpcom_launchpad_update_task_status()`, we should only mark tasks as complete when we have a defined task for the mapped task ID:
```php
		if ( ! isset( $task_definitions[ $actual_task_id ] ) ) {
			// Bail if no task definition was found.
			continue;
		}
```

In the test setup, there is no task with an ID of `THIS_IS_THE_MAP`, so I am confused as to how this works or should work. Should we have separate test cases to handle the following situations:
 1. Task is mapped to a valid task ID -- we should mark the underlying task as complete
 2. Task is mapped to an invalid task ID -- we should _not_ mark the underlying task as complete

Reply:
Depending on the outcome of [this conversation](https://github.com/Automattic/jetpack/pull/31471#discussion_r1285871674) this might change.

Reply:
I am going to mark this conversation as resolved. We don't need a task for the `id_map` to work.